core/drivers/crypto/se050/core/cipher.c

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2020 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <se050_cipher_algorithms.h>
+#include <crypto/crypto_impl.h>
+#include <drvcrypt.h>
+#include <drvcrypt_cipher.h>
+#include <initcall.h>
+#include <string.h>
+#include <utee_defines.h>
+#include <util.h>
+
+static TEE_Result do_init(struct drvcrypt_cipher_init *dinit)
+{
+	struct crypto_cipher_ctx *ctx = dinit->ctx;
+
+	return ctx->ops->init(dinit->ctx, dinit->encrypt,
+			      dinit->key1.data, dinit->key1.length,
+			      dinit->key2.data, dinit->key2.length,
+			      dinit->iv.data, dinit->iv.length);
+}
+
+static TEE_Result do_update(struct drvcrypt_cipher_update *dupdate)
+{
+	struct crypto_cipher_ctx *ctx = dupdate->ctx;
+
+	return ctx->ops->update(ctx, dupdate->last, dupdate->src.data,
+				dupdate->src.length, dupdate->dst.data);
+}
+
+static void do_final(void *context)
+{
+	struct crypto_cipher_ctx *ctx = context;
+
+	ctx->ops->final(ctx);
+}
+
+static void do_free(void *context)
+{
+	struct crypto_cipher_ctx *ctx = context;
+
+	ctx->ops->free_ctx(ctx);
+}
+
+static void do_copy_state(void *out, void *in)
+{
+	struct crypto_cipher_ctx *dst_ctx = out;
+	struct crypto_cipher_ctx *src_ctx = in;
+
+	src_ctx->ops->copy_state(dst_ctx, src_ctx);
+}
+
+static TEE_Result do_allocate(void **ctx, uint32_t algo)
+{
+	TEE_Result res = TEE_ERROR_NOT_IMPLEMENTED;
+
+	switch (algo) {
+	case TEE_ALG_AES_CTR:
+		res = se050_aes_ctr_allocate(ctx);
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	return res;
+}
+
+static struct drvcrypt_cipher driver_cipher = {
```

Comment:
ywa I already have support for all the other modes...I just havent posted because this PR is messy as it is. once this one is in the rearview mirror (if ever)  I'll start posting the rest....